xarray/core/dataset.py

```diff
@@ -4071,8 +4072,55 @@ def to_dataframe(self):
         """
         return self._to_dataframe(self.dims)
 
+    def _set_sparse_data_from_dataframe(self, dataframe, dims, shape):
+        from sparse import COO
+
+        idx = dataframe.index
+        if isinstance(idx, pd.MultiIndex):
+            try:
+                codes = idx.codes
+            except AttributeError:
+                # deprecated since pandas 0.24
+                codes = idx.labels
+            coords = np.stack([np.asarray(code) for code in codes], axis=0)
```

Comment:
You can change the whole paragraph to
```python
            try:
                coords = np.asarray(idx.codes)
            except AttributeError:
                # deprecated since pandas 0.24
                coords = np.asarray(idx.labels)

Reply:
Pandas uses a weird `FrozenList` type for codes. I'd rather explicitly iterate over it than rely upon NumPy properly converting the whole thing at once.